plugins/node/opentelemetry-hapi-instrumentation/src/hapi.ts

```diff
@@ -0,0 +1,373 @@
+          }
+        }
+      } else if (args.length == 1) {
+        const eventObj:
+          | Hapi.ServerExtEventsObject
+          | Hapi.ServerExtEventsRequestObject = args[0];
+        if (isLifecycleExtType(eventObj.type)) {
+          const lifecycleEventObj = eventObj as Hapi.ServerExtEventsRequestObject;
+          const handler = instrumentation._wrapExtMethods(
+            lifecycleEventObj.method,
+            eventObj.type,
+            pluginName
+          );
+          lifecycleEventObj.method = handler;
+          args[0] = lifecycleEventObj;
+        }
+      } else {
+        const type = args[0];
+        if (isLifecycleExtType(type)) {
+          const method: PatchableExtMethod = args[1];
+          args[1] = instrumentation._wrapExtMethods(method, type, pluginName);
+        }
+      }
+      return original.apply(this, args);
+    };
+  }
+
+  /**
+   * Patches the Server.route function. This function accepts either one or an array
+   * of Hapi.ServerRoute objects and adds instrumentation on each route via a call to
+   * the @function _wrapRouteHandler function.
+   * @param {HapiServerRouteInputMethod} original - the original route function which adds
+   * the route to the server
+   * @param {string} [pluginName] - if present, represents the name of the plugin responsible
+   * for adding this server route. Else, signifies that the route was added directly
+   */
+  private _getServerRoutePatch(
+    original: HapiServerRouteInputMethod,
+    pluginName?: string
+  ) {
+    const instrumentation = this;
+    this._logger.debug('Patching Hapi.Server route function');
+    return function route(
+      this: Hapi.Server,
+      route: HapiServerRouteInput
+    ): void {
+      if (Array.isArray(route)) {
+        for (let i = 0; i < route.length; i++) {
+          const newRoute = instrumentation._wrapRouteHandler.call(
+            instrumentation,
+            route[i],
+            pluginName
+          );
+          route[i] = newRoute;
+        }
+      } else {
+        route = instrumentation._wrapRouteHandler.call(
+          instrumentation,
+          route,
+          pluginName
+        );
+      }
+      return original.apply(this, [route]);
+    };
+  }
+
+  /**
+   * Wraps newly registered plugins to add instrumentation to the plugin's clone of
+   * the original server. Specifically, wraps the server.route and server.ext functions
+   * via calls to @function _getServerRoutePatch and @function _getServerExtPatch
+   * @param {Hapi.Plugin<T>} plugin - the new plugin which is being instrumented
+   */
+  private _wrapRegisterHandler<T>(plugin: Hapi.Plugin<T>): void {
+    const instrumentation = this;
+    const pluginName = getPluginName(plugin);
+    const oldHandler = plugin.register;
+    const newRegisterHandler = function (server: Hapi.Server, options: T) {
+      shimmer.wrap(server, 'route', original => {
+        return instrumentation._getServerRoutePatch.bind(instrumentation)(
```

Comment:
I'm wondering why there's the need to bind the functions here since you're immediately invoking them here. Wouldn't the method syntax be sufficient here?

Reply:
In my understanding, the functions need to be bound because shimmer is wrapping the `server.route` function so that this function is not invoked until `server.route` is actually called by the user. It would also be possible to just pass in `this` as an extra parameter (named instrumentation), but after discussion with the OTel contributors on my Koa PR, we determined that for future readability it would be better for "this" to continue referring to my instrumentation